### PR TITLE
Simple czm laws without damage

### LINKED ISSUE #18: Add simple no damage cohesive laws
We need to have some simple cohesive law either for testing models or for including a simple interface behavior .

### PR DESCRIPTION
**Pull request by arovinelli**
This PR adds two simple no-damage cohesive laws:
1.  a complete linear elastic law
2.  a linear elastic cohesive law with viscous sliding (normal behavior is still linear elastic)

These classes are the basis to develop more complex laws, for instance the sham-needleman GB cavitation model can be derived starting from the viscous law

ref #19 
close #18

### CODE REVIEW

File: src/materials/czm/PureElasticCZM.C [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "InterfaceValueTools.h"
+#include "PureElasticCZM.h"
+
+registerMooseObject("DeerApp", PureElasticCZM);
+
+template <> InputParameters validParams<PureElasticCZM>() {
```
Comment (reverendbedford) [inline src/materials/czm/PureElasticCZM.C]: Capitalize description sentence.

File: src/materials/czm/PureElasticCZM.C [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  ComputeNormalTractionDerivatives(traction_jump_derivatives);
+  ComputeShearTractionDerivatives(traction_jump_derivatives);
+
+  return traction_jump_derivatives;
+}
+
+void PureElasticCZM::ComputeShearTraction(RealVectorValue &traction) {
+  Real C = ComputeShearStiffness();
+  for (unsigned int i = 1; i < 3; i++)
```
Comment (reverendbedford) [inline src/materials/czm/PureElasticCZM.C]: Here and elsewhere: this should not be a magic "3".  What happens in 2D?  Or at the very least you should be throwing an error for 2D meshes.

See the moose tensor_mechanics kernels for how they deal with  it.

File: src/materials/czm/ViscousSlidingCZM.C [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "InterfaceValueTools.h"
+#include "RotationMatrix.h"
+#include "ViscousSlidingCZM.h"
+
+registerMooseObject("DeerApp", ViscousSlidingCZM);
```
Comment (reverendbedford) [inline src/materials/czm/ViscousSlidingCZM.C]: Capitalize helptext.

File: src/materials/czm/ViscousSlidingCZM.C [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+void ViscousSlidingCZM::computeQpProperties() {
+
+  RealTensorValue RotationGlobalToLocal =
+      RotationMatrix::rotVec1ToVec2(_normals[_qp], RealVectorValue(1, 0, 0));
+
+  // computing the displacement jump rate
+  for (unsigned int i = 0; i < _ndisp; i++)
+    _displacement_jump_global_dot[_qp](i) =
+        (*_disp_neighbor_dot[i])[_qp] - (*_disp_dot[i])[_qp];
+  for (unsigned int i = _ndisp; i < 3; i++)
```
Comment (reverendbedford) [inline src/materials/czm/ViscousSlidingCZM.C]: Hmm, so maybe I don't understand what you're doing above.

File: src/materials/czm/ViscousSlidingCZM.C [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void ViscousSlidingCZM::ComputeShearTractionDerivatives(
+    RankTwoTensor &traction_derivatives) {
+  Real C = ComputeShearStiffness();
+  RankTwoTensor dC_dui = ComputeShearStiffnessDerivatives();
+  Real eta = ComputeShearViscosity();
+  RankTwoTensor deta_dui = ComputeShearViscosityDerivatives();
+  for (unsigned int i = 1; i < 3; i++) {
+
```
Comment (reverendbedford) [inline src/materials/czm/ViscousSlidingCZM.C]: Blarg whitespace

File: src/materials/czm/ViscousSlidingCZM.C [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    traction_derivatives(i, i) =
+        dTsi_dui + dTsi_dC * dC_dui(i, i) + dTsi_deta * deta_dui(i, i);
+  }
+}
+
+Real ViscousSlidingCZM::ComputeShearViscosity() { return _shear_viscosity; };
+
+RankTwoTensor ViscousSlidingCZM::ComputeShearViscosityDerivatives() {
+  RankTwoTensor deta_dui;
```
Comment (reverendbedford) [inline src/materials/czm/ViscousSlidingCZM.C]: So this is the zero tensor?  There's a way to get that as a static method of the RankTwoTensor class that would make it much more clear.

File: include/materials/czm/PureElasticCZM.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+class PureElasticCZM : public CZMMaterialBase {
+public:
+  PureElasticCZM(const InputParameters &parameters);
+
+protected:
+  virtual RealVectorValue computeTraction() override;
+
+  virtual RankTwoTensor computeTractionDerivatives() override;
+
+  /// the displacement jump associated to the maximum traction
```
[the remark this answers is not in the document]
Comment (arovinelli) [reply include/materials/czm/PureElasticCZM.h]: I'm doing this

File: src/materials/czm/PureElasticCZM.C (hunk 2 above)
Comment (arovinelli) [reply src/materials/czm/PureElasticCZM.C]: This is a magic 3 :grin: as the interface model always work in 3D even for 2D meshes.
I'll add a comment but is also already documented in MOOSE in the CZMMaterialBase class

**Pull request merged**